### PR TITLE
Implement sharing for hls-graph Keys

### PR DESCRIPTION
**Pull request by wz1000**
On large projects, duplicate Keys start to comprise a significant (~5-10%) fraction of the heap.
This becomes better if we use `Int`s as keys and de-duplicate the actual values by storing them
in maps. `Key`s are mostly just compared and hash during usual operation, printing them out is
rare so a lookup shouldn't pessimize it much.

Also use HashSets to keep track of dependencies.
However, this means that the underlying value for `Key`s will never be garbage collected once allocated.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3206"></a>

### CODE REVIEW

File: hls-graph/src/Development/IDE/Graph/Internal/Types.hs [hunk 1]
```diff
@@ -84,16 +89,54 @@ data ShakeDatabase = ShakeDatabase !Int [Action ()] Database
...
+instance Hashable Key where
+  hashWithSalt i (UnsafeMkKey x) = hashWithSalt i x
 instance Show Key where
-    show (Key x) = show x
+  show (Key x) = show x
+
+instance Eq KeyValue where
+    KeyValue a == KeyValue b = Just a == cast b
+instance Hashable KeyValue where
+    hashWithSalt i (KeyValue x) = hashWithSalt i (typeOf x, x)
```
[the remark this answers is not in the document]
Comment (wz1000) [reply hls-graph/src/Development/IDE/Graph/Internal/Types.hs:137]: How can you hash the key, which is derived from the order of insertions into the map, from the `KeyValue`?

[the remark this answers is not in the document]
Comment (pepeiborra) [reply hls-graph/src/Development/IDE/Graph/Internal/Types.hs:137]: Can't you store the `Key` itself in the `KeyValue`?

[the remark this answers is not in the document]
Comment (wz1000) [reply hls-graph/src/Development/IDE/Graph/Internal/Types.hs:137]: But you need the `KeyValue` first to check in the map for duplicates.

[the remark this answers is not in the document]
Comment (pepeiborra) [reply hls-graph/src/Development/IDE/Graph/Internal/Types.hs:137]: Right

**Review by pepeiborra — APPROVED** [review]
Approved. Please share some benchmarks showing the memory savings when you have a chance!

**Review by pepeiborra — CHANGES_REQUESTED** [review]
Can we drop the `Int` proxy (`Key`) and the `IntMap`? 
`newKey` will lookup in the `KeyMap` and return the existing value if any, recovering sharing. 

What's the benefit of keeping an extra mapping?

**Comment by wz1000** [comment]
> Can we drop the `Int` proxy (`Key`) and the `IntMap`? `newKey` will lookup in the `KeyMap` and return the existing value if any, recovering sharing.
> 
> What's the benefit of keeping an extra mapping?

I tried this first, but GHC unpacks the Key and we lose sharing at that point.

**Comment by pepeiborra** [comment]
> I tried this first, but GHC unpacks the Key and we lose sharing at that point.

Worth adding a comment about this, I believe my previous attempt to share keys must have failed for this reason

**Comment by michaelpj** [comment]
> I tried this first, but GHC unpacks the Key and we lose sharing at that point.

Maybe it's a bit too sensitive, but can we do something to prevent GHC from unpacking the key? e.g. not import the constructor in the module in question?

File: hls-graph/src/Development/IDE/Graph/Internal/Types.hs [hunk 2]
```diff
@@ -84,16 +89,54 @@ data ShakeDatabase = ShakeDatabase !Int [Action ()] Database
...
+lookupKeyValue (UnsafeMkKey x) = unsafePerformIO $ do
+  KeyMap _ im _ <- readIORef keyMap
+  pure $! fromJust (IM.lookup x im)
+
+{-# NOINLINE lookupKeyValue #-}
+
+instance Eq Key where
+  UnsafeMkKey a == UnsafeMkKey b = a == b
+instance Hashable Key where
+  hashWithSalt i (UnsafeMkKey x) = hashWithSalt i x
```
[the remark this answers is not in the document]
Comment (wz1000) [reply hls-graph/src/Development/IDE/Graph/Internal/Types.hs:130]: I think it makes it clearer to write these explicitly, given that we have some weird stuff going on with the Key pattern synonym.

File: hls-graph/src/Development/IDE/Graph/Internal/Types.hs [hunk 3]
```diff
@@ -149,24 +192,24 @@ data Result = Result {
     resultData      :: !BS.ByteString
     }
 
-data ResultDeps = UnknownDeps | AlwaysRerunDeps ![Key] | ResultDeps ![Key]
+data ResultDeps = UnknownDeps | AlwaysRerunDeps !(HashSet Key) | ResultDeps !(HashSet Key)
```
[the remark this answers is not in the document]
Comment (wz1000) [reply hls-graph/src/Development/IDE/Graph/Internal/Types.hs]: I've introduced `KeyMap` and `KeySet` backed by `IntMap` and `IntSet`.

**Comment by wz1000** [comment]
> Maybe it's a bit too sensitive, but can we do something to prevent GHC from unpacking the key? e.g. not import the constructor in the module in question?

I don't think this is possible currently. Would be a good feature request though.

**Comment by wz1000** [comment]
I've also added sharing of the rendered `Key` as `Text` to avoid duplicating the rendered Key in the diagnostic store, else we get a duplicate of each key as a rendered `Text` value for each file in the project.

**Review by pepeiborra — APPROVED** [review]
Can we have those benchmark results?

**Comment by wz1000** [comment]
Here are some profiles from before and after

before (heap size): 
2022-10-21-211021_grab

before (keys in heap): 
2022-10-21-211050_grab

In the second image you can see that just the top allocation site of duplicate keys is responsible for ~200MB of live bytes. These are all eliminated by this patch.

after (heap size): 
2022-09-27-140932_grab

**Pull request merged**